gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/EntityService.java

```diff
@@ -235,6 +246,61 @@ public BulkUpdateEntitiesResponse bulkUpdateEntities(
     return edsEntityUpdater.bulkUpdateEntities(request, updateExecutionContext);
   }
 
+  private EntitiesRequest addEntityIdsFromInteractionFilter(
+      RequestContext requestContext, EntitiesRequest preProcessedRequest) {
+    Set<EntityKey> entityKeys =
+        this.interactionsFetcher.fetchInteractionsIdsIfNecessary(
+            requestContext, preProcessedRequest);
+    if (!entityKeys.isEmpty()) {
+      Filter entityIdsInFilter =
+          createEntityIdInFilter(requestContext, preProcessedRequest.getEntityType(), entityKeys);
+
+      EntitiesRequest.Builder preProcessRequestBuilder = preProcessedRequest.toBuilder();
+      if (preProcessRequestBuilder.getFilter().equals(Filter.getDefaultInstance())) {
+        preProcessRequestBuilder.setFilter(entityIdsInFilter);
+      } else {
+        preProcessRequestBuilder.setFilter(
+            Filter.newBuilder()
+                .setOperator(Operator.AND)
+                .addChildFilter(preProcessRequestBuilder.getFilter())
+                .addChildFilter(entityIdsInFilter)
+                .build());
+      }
+      preProcessedRequest = preProcessRequestBuilder.build();
+    }
+    return preProcessedRequest;
+  }
+
+  private Filter createEntityIdInFilter(
+      RequestContext requestContext, String entityType, Set<EntityKey> entityKeys) {
+    List<EntityKey> entities = new ArrayList<>(entityKeys);
+    if (entities.size() > DEFAULT_QUERY_SERVICE_GROUP_BY_LIMIT) {
+      entities = entities.subList(0, DEFAULT_QUERY_SERVICE_GROUP_BY_LIMIT);
+    }
+    List<String> entityIds =
+        entities.stream().map(EntityKey::toString).collect(Collectors.toUnmodifiableList());
+
+    List<String> idAttributeIds =
+        AttributeMetadataUtil.getIdAttributeIds(
+            metadataProvider, entityIdColumnsConfigs, requestContext, entityType);
+
+    List<Filter> entityIdsInFilter =
```

Comment:
A bit confused. If we have multiple id attributes for an entity, why are we adding the same id attribute values for all the multiple id attributes

Reply:
Yeah this looks like a bug to me. If anything is using multiple ID attributes (not sure if that's the case today), this would fail because we're assuming each one should contain the same value, while the intention was that each part contains a portion of the ID.

Reply:
> I assumed the id filter evaluation based on here org.hypertrace.gateway.service.common.datafetcher.EntityInteractionsFetcher#createFilterForEntityKeys. Ideally this should never happen. No one is using multi id attributes but i believe this is legacy code written in that way

Threading the result here since lots of comments.

If this should never happen, then we should ideally fix gateway service to remove support - or fix the code to support it, although not sure how that would work. If we can't do either, at a minimum make the code loudly error out if it ever runs into this scenario. Right now, we're making the code support it, but the logic doesn't - so if it were ever used it would lead to very annoying to track down bugs.